pkg/machine/compression/sparse_file_writer.go

```diff
@@ -0,0 +1,135 @@
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"os"
+)
+
+type state int
+
+const (
+	zerosThreshold = 1024
+
+	stateData = iota
+	stateZeros
+)
+
+type SparseWriterUnderlyingFile interface {
+	Seek(offset int64, whence int) (ret int64, err error)
+	Close() error
+	Write(b []byte) (n int, err error)
+}
+
+type SparseWriter struct {
+	state      state
+	file       SparseWriterUnderlyingFile
+	zeros      int64
+	lastIsZero bool
+}
+
+func newSparseWriterToFile(file SparseWriterUnderlyingFile) *SparseWriter {
+	return &SparseWriter{file: file, state: stateData}
+}
+
+func NewSparseWriter(filePath string) (*SparseWriter, error) {
+	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666)
+	if err != nil {
+		return nil, err
+	}
+	return &SparseWriter{file: file, state: stateData}, nil
+}
+
+func (sw *SparseWriter) createHole() error {
+	zeros := sw.zeros
+	if zeros == 0 {
+		return nil
+	}
+	sw.zeros = 0
+	sw.lastIsZero = true
+	_, err := sw.file.Seek(zeros, io.SeekCurrent)
+	return err
+}
+
+func findFirstNotZero(b []byte) int {
+	for i, v := range b {
+		if v != 0 {
+			return i
+		}
+	}
+	return -1
+}
+
+// Write writes data to the file, creating holes for long sequences of zeros.
+func (sw *SparseWriter) Write(data []byte) (int, error) {
+	written, current := 0, 0
+	totalLen := len(data)
+	for current < len(data) {
+		switch sw.state {
+		case stateData:
+			nextZero := bytes.IndexByte(data[current:], 0)
+			if nextZero < 0 {
+				_, err := sw.file.Write(data[written:])
+				sw.lastIsZero = false
+				return totalLen, err
+			} else {
+				current += nextZero
+				sw.state = stateZeros
+			}
+		case stateZeros:
+			nextNonZero := findFirstNotZero(data[current:])
+			if nextNonZero < 0 {
+				// finish with a zero, flush any data and keep track of the zeros
+				if written != current {
+					if _, err := sw.file.Write(data[written:current]); err != nil {
+						return -1, err
+					}
+					sw.lastIsZero = false
+				}
+				sw.zeros += int64(len(data) - current)
+				return totalLen, nil
+			}
+			// do not bother with too short sequences
+			if sw.zeros == 0 && nextNonZero < zerosThreshold {
+				sw.state = stateData
+				current += nextNonZero
+				continue
+			}
+			if written != current {
+				if _, err := sw.file.Write(data[written:current]); err != nil {
+					return -1, err
+				}
+				sw.lastIsZero = false
+			}
+			sw.zeros += int64(nextNonZero)
+			current += nextNonZero
+			if err := sw.createHole(); err != nil {
+				return -1, err
+			}
+			written = current
+		}
+	}
+	return totalLen, nil
+}
+
+// Close closes the SparseWriter's underlying file.
+func (sw *SparseWriter) Close() error {
+	if sw.file == nil {
+		return errors.New("file is already closed")
+	}
+	if err := sw.createHole(); err != nil {
+		return err
+	}
+	if sw.lastIsZero {
+		if _, err := sw.file.Seek(-1, os.SEEK_CUR); err != nil {
+			return err
+		}
+		if _, err := sw.file.Write([]byte{0}); err != nil {
+			return err
+		}
+	}
+	err := sw.file.Close()
```

Comment:
Should this happen, even if there is an error?

Reply:
yes, I'll fix it